sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/datastore/DataStoreV1Table.java

```diff
@@ -0,0 +1,483 @@
+      private Key constructKeyFromRow(Row row) {
+        if (!row.getSchema().getFieldNames().contains(keyField)) {
+          // When key field is not present - use key supplier to generate a random one.
+          return makeKey(kind, keySupplier.get()).build();
+        }
+        byte[] keyBytes = row.getBytes(keyField);
+        try {
+          return Key.parseFrom(keyBytes);
+        } catch (InvalidProtocolBufferException e) {
+          throw new IllegalStateException("Failed to parse DataStore key from bytes.");
+        }
+      }
+
+      /**
+       * Converts a {@code Row} value to an appropriate DataStore {@code Value} object.
+       *
+       * @param value {@code Row} value to convert.
+       * @throws IllegalStateException when no mapping function for object of given type exists.
+       * @return resulting {@code Value}.
+       */
+      private Value mapObjectToValue(Object value) {
+        if (value == null) {
+          return Value.newBuilder().build();
+        }
+
+        if (Boolean.class.equals(value.getClass())) {
+          return makeValue((Boolean) value).build();
+        } else if (Byte.class.equals(value.getClass())) {
+          return makeValue((Byte) value).build();
+        } else if (Long.class.equals(value.getClass())) {
+          return makeValue((Long) value).build();
+        } else if (Short.class.equals(value.getClass())) {
+          return makeValue((Short) value).build();
+        } else if (Integer.class.equals(value.getClass())) {
+          return makeValue((Integer) value).build();
+        } else if (Double.class.equals(value.getClass())) {
+          return makeValue((Double) value).build();
+        } else if (Float.class.equals(value.getClass())) {
+          return makeValue((Float) value).build();
+        } else if (String.class.equals(value.getClass())) {
+          return makeValue((String) value).build();
+        } else if (Instant.class.equals(value.getClass())) {
+          return makeValue(((Instant) value).toDate()).build();
+        } else if (byte[].class.equals(value.getClass())) {
+          return makeValue(ByteString.copyFrom((byte[]) value)).build();
+        } else if (value instanceof Row) {
+          // Recursive conversion to handle nested rows.
+          Row row = (Row) value;
+          return makeValue(constructEntityFromRow(row.getSchema(), row)).build();
+        } else if (value instanceof Collection) {
+          // Recursive to handle nested collections.
+          Collection<Object> collection = (Collection<Object>) value;
+          List<Value> arrayValues =
+              collection.stream().map(this::mapObjectToValue).collect(Collectors.toList());
+          return makeValue(arrayValues).build();
+        }
+        throw new IllegalStateException(
+            "No conversion exists from type: "
+                + value.getClass()
+                + " to DataStove Value. Supported types are: "
+                + SUPPORTED_JAVA_TYPES.toString());
```

Comment:
This list looks to exist only for this error message (which appears to be an internal error), this is going to get out of sync. You might want to just remove it.

Reply:
Done.